File: generation/dnp3/src/main/scala/com/automatak/render/dnp3/objects/groups/Group42.scala
/**
 * Copyright 2013-2020 Automatak, LLC
 *
 * Licensed to Green Energy Corp (www.greenenergycorp.com) and Automatak
 * LLC (www.automatak.com) under one or more contributor license agreements.
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership. Green Energy Corp and Automatak LLC license
 * this file to you under the Apache License, Version 2.0 (the "License"); you
 * may not use this file except in compliance with the License. You may obtain
 * a copy of the License at:
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.automatak.render.dnp3.objects.groups

import com.automatak.render.dnp3.objects._

import FixedSizeField._
import com.automatak.render.dnp3.objects.VariationNames._
import com.automatak.render.dnp3.objects.generators.ConversionToAnalogOutputStatus

//analog output events
object Group42 extends ObjectGroup {
  def objects = List( Group42Var0, Group42Var1, Group42Var2, Group42Var3, Group42Var4, Group42Var5, Group42Var6, Group42Var7, Group42Var8)
  def group: Byte = 42
  def desc: String = "Analog Output Event"
  def isEventGroup: Boolean = true
}

object Group42Var0 extends AnyVariation(Group42, 0)
object Group42Var1 extends FixedSize(Group42, 1, bit32WithFlag)(flags, value32) with ConversionToAnalogOutputStatus
object Group42Var2 extends FixedSize(Group42, 2, bit16WithFlag)(flags, value16) with ConversionToAnalogOutputStatus
object Group42Var3 extends FixedSize(Group42, 3, bit32WithFlagTime)(flags, value32, time48) with ConversionToAnalogOutputStatus
object Group42Var4 extends FixedSize(Group42, 4, bit16WithFlagTime)(flags, value16, time48) with ConversionToAnalogOutputStatus
object Group42Var5 extends FixedSize(Group42, 5, singlePrecisionWithFlag)(flags, float32) with ConversionToAnalogOutputStatus
object Group42Var6 extends FixedSize(Group42, 6, doublePrecisionWithFlag)(flags, float64) with ConversionToAnalogOutputStatus
object Group42Var7 extends FixedSize(Group42, 7, singlePrecisionWithFlagTime)(flags, float32, time48) with ConversionToAnalogOutputStatus
object Group42Var8 extends FixedSize(Group42, 8, doublePrecisionWithFlagTime)(flags, float64, time48) with ConversionToAnalogOutputStatus

File: cpp/opendnp3/src/opendnp3/objects/Group42.cpp
//
//  _   _         ______    _ _ _   _             _ _ _
// | \ | |       |  ____|  | (_) | (_)           | | | |
// |  \| | ___   | |__   __| |_| |_ _ _ __   __ _| | | |
// | . ` |/ _ \  |  __| / _` | | __| | '_ \ / _` | | | |
// | |\  | (_) | | |___| (_| | | |_| | | | | (_| |_|_|_|
// |_| \_|\___/  |______\__,_|_|\__|_|_| |_|\__, (_|_|_)
//                                           __/ |
//                                          |___/
// Copyright 2013 Automatak LLC
// 
// Automatak LLC (www.automatak.com) licenses this file
// to you under the the Apache License Version 2.0 (the "License"):
// 
// http://www.apache.org/licenses/LICENSE-2.0.html
//

#include "Group42.h"

#include <openpal/Serialization.h>

using namespace openpal;

namespace opendnp3 {

const GroupVariationID  Group42Var1::ID(42,1);

Group42Var1 Group42Var1::Read(ReadOnlyBuffer& buffer)
{
  Group42Var1 obj;
  obj.flags = UInt8::Read(buffer);
  buffer.Advance(1);
  obj.value = Int32::Read(buffer);
  buffer.Advance(4);
  return obj;
}

void Group42Var1::Write(const Group42Var1& arg, openpal::WriteBuffer& buffer)
{
  UInt8::Write(buffer, arg.flags);
  buffer.Advance(1);
  Int32::Write(buffer, arg.value);
  buffer.Advance(4);
}

const GroupVariationID  Group42Var2::ID(42,2);

Group42Var2 Group42Var2::Read(ReadOnlyBuffer& buffer)
{
  Group42Var2 obj;
  obj.flags = UInt8::Read(buffer);
  buffer.Advance(1);
  obj.value = Int16::Read(buffer);
  buffer.Advance(2);
  return obj;
}

void Group42Var2::Write(const Group42Var2& arg, openpal::WriteBuffer& buffer)
{
  UInt8::Write(buffer, arg.flags);
  buffer.Advance(1);
  Int16::Write(buffer, arg.value);
  buffer.Advance(2);
}

const GroupVariationID  Group42Var3::ID(42,3);

Group42Var3 Group42Var3::Read(ReadOnlyBuffer& buffer)
{
  Group42Var3 obj;
  obj.flags = UInt8::Read(buffer);
  buffer.Advance(1);
  obj.value = Int32::Read(buffer);
  buffer.Advance(4);
  obj.time = UInt48::Read(buffer);
  buffer.Advance(6);
  return obj;
}

void Group42Var3::Write(const Group42Var3& arg, openpal::WriteBuffer& buffer)
{
  UInt8::Write(buffer, arg.flags);
  buffer.Advance(1);
  Int32::Write(buffer, arg.value);
  buffer.Advance(4);
  UInt48::Write(buffer, arg.time);
  buffer.Advance(6);
}

const GroupVariationID  Group42Var4::ID(42,4);

Group42Var4 Group42Var4::Read(ReadOnlyBuffer& buffer)
{
  Group42Var4 obj;
  obj.flags = UInt8::Read(buffer);
  buffer.Advance(1);
  obj.value = Int16::Read(buffer);
  buffer.Advance(2);
  obj.time = UInt48::Read(buffer);
  buffer.Advance(6);
  return obj;
}

void Group42Var4::Write(const Group42Var4& arg, openpal::WriteBuffer& buffer)
{
  UInt8::Write(buffer, arg.flags);
  buffer.Advance(1);
  Int16::Write(buffer, arg.value);
  buffer.Advance(2);
  UInt48::Write(buffer, arg.time);
  buffer.Advance(6);
}

const GroupVariationID  Group42Var5::ID(42,5);

Group42Var5 Group42Var5::Read(ReadOnlyBuffer& buffer)
{
  Group42Var5 obj;
  obj.flags = UInt8::Read(buffer);
  buffer.Advance(1);
  obj.value = SingleFloat::Read(buffer);
  buffer.Advance(4);
  return obj;
}

void Group42Var5::Write(const Group42Var5& arg, openpal::WriteBuffer& buffer)
{
  UInt8::Write(buffer, arg.flags);
  buffer.Advance(1);
  SingleFloat::Write(buffer, arg.value);
  buffer.Advance(4);
}

const GroupVariationID  Group42Var6::ID(42,6);

Group42Var6 Group42Var6::Read(ReadOnlyBuffer& buffer)
{
  Group42Var6 obj;
  obj.flags = UInt8::Read(buffer);
  buffer.Advance(1);
  obj.value = DoubleFloat::Read(buffer);
  buffer.Advance(8);
  return obj;
}

void Group42Var6::Write(const Group42Var6& arg, openpal::WriteBuffer& buffer)
{
  UInt8::Write(buffer, arg.flags);
  buffer.Advance(1);
  DoubleFloat::Write(buffer, arg.value);
  buffer.Advance(8);
}

const GroupVariationID  Group42Var7::ID(42,7);

Group42Var7 Group42Var7::Read(ReadOnlyBuffer& buffer)
{
  Group42Var7 obj;
  obj.flags = UInt8::Read(buffer);
  buffer.Advance(1);
  obj.value = SingleFloat::Read(buffer);
  buffer.Advance(4);
  obj.time = UInt48::Read(buffer);
  buffer.Advance(6);
  return obj;
}

void Group42Var7::Write(const Group42Var7& arg, openpal::WriteBuffer& buffer)
{
  UInt8::Write(buffer, arg.flags);
  buffer.Advance(1);
  SingleFloat::Write(buffer, arg.value);
  buffer.Advance(4);
  UInt48::Write(buffer, arg.time);
  buffer.Advance(6);
}

const GroupVariationID  Group42Var8::ID(42,8);

Group42Var8 Group42Var8::Read(ReadOnlyBuffer& buffer)
{
  Group42Var8 obj;
  obj.flags = UInt8::Read(buffer);
  buffer.Advance(1);
  obj.value = DoubleFloat::Read(buffer);
  buffer.Advance(8);
  obj.time = UInt48::Read(buffer);
  buffer.Advance(6);
  return obj;
}

void Group42Var8::Write(const Group42Var8& arg, openpal::WriteBuffer& buffer)
{
  UInt8::Write(buffer, arg.flags);
  buffer.Advance(1);
  DoubleFloat::Write(buffer, arg.value);
  buffer.Advance(8);
  UInt48::Write(buffer, arg.time);
  buffer.Advance(6);
}


}
